File: packages/wire/frame.go
package wire

import (
	"encoding/binary"
	"fmt"
)

const frameHeaderBytes = 16

// FrameVersion is the frame-header format version (the header's first byte). It mirrors
// FRAME_VERSION in packages/protocol/src/constants.ts and is bumped only if the header
// layout changes.
const FrameVersion uint8 = 1

// FrameHeader flags (the Flags byte).
const (
	// FrameFlagLastInBlock marks the final data frame in a logical block.
	FrameFlagLastInBlock uint8 = 0x01
	// FrameFlagPadded indicates the AEAD payload is padded to a power-of-two bucket (V17-PR03).
	FrameFlagPadded uint8 = 0x02
	// FlagPadded is an alias for FrameFlagPadded.
	FlagPadded uint8 = FrameFlagPadded
)

// Padding parameters.
const (
	// PaddingCapability is the feature announced in caps for negotiated traffic padding.
	PaddingCapability = "padding"
	// MinPadBucketSize is the smallest quantized bucket (256 bytes).
	MinPadBucketSize = 256
	// MaxPadBucketSize is the maximum frame bucket matching u16 max (65535 bytes).
	MaxPadBucketSize = 65535
)

// FrameHeader is the 16-byte frame header. Its encoded bytes are the
// AES-GCM additional authenticated data, so the codec must be exact and stable.
//
// Layout (big-endian):
//
//	version(u8) type(u8) flags(u8) reserved(u8)
//	fileIdx(u16) blockIdx(u32) frameOff(u32) len(u16)
//
// FrameOff is a byte offset within the block, so it must span the whole block: at the
// default 1 MiB block a 16 KiB frame reaches offset ~1 MiB, far past u16. It is u32
// (not u16, as an earlier draft miscalculated), which the header absorbs by dropping the
// former trailing reserved(u16) — the header stays a fixed 16 bytes.
type FrameHeader struct {
	Version  uint8
	Type     uint8
	Flags    uint8
	FileIdx  uint16
	BlockIdx uint32
	FrameOff uint32 // byte offset within the block, not the file
	Len      uint16 // ciphertext payload length
}

// Frame type tags (the Type byte). Mirrors FrameType in transfer.ts.
const (
	FrameCaps        uint8 = 1
	FrameManifest    uint8 = 2
	FrameBlockData   uint8 = 3
	FrameBlockHash   uint8 = 4
	FrameBlockRecv   uint8 = 5
	FrameAck         uint8 = 6
	FrameNack        uint8 = 7
	FrameControl     uint8 = 8
	FrameComplete    uint8 = 9
	FrameDone        uint8 = 10
	FrameFail        uint8 = 11
	FrameResumeState uint8 = 12
	// FrameResumeAuth carries one resume-auth message (EncodeResumeMessage JSON) sealed
	// under the SESSION directional keys, exchanged strictly before the transfer engine
	// starts (V13-PR08). It is never used for transfer protocol frames; after mutual
	// authentication the transfer runs under the fresh resumed key epoch.
	FrameResumeAuth uint8 = 13
	// FramePairingExchange carries one pairing message (PairingRequest, PairingResponse,
	// or PairingConfirm JSON) sealed under the SESSION directional keys (V15-PR02).
	FramePairingExchange uint8 = 14
	// FrameTrustedAuth carries one trusted-session handshake message (TrustedAuthInit,
	// TrustedAuthResponse, or TrustedAuthConfirm JSON) for paired devices (V15-PR03).
	FrameTrustedAuth uint8 = 15
)

// PadBucketSize returns the smallest power-of-two bucket size >= (unpaddedLen + 2),
// clamped to [MinPadBucketSize, MaxPadBucketSize].
func PadBucketSize(unpaddedLen int) int {
	target := unpaddedLen + 2
	if target <= MinPadBucketSize {
		return MinPadBucketSize
	}
	if target > 32768 {
		return MaxPadBucketSize
	}
	bucket := MinPadBucketSize
	for bucket < target {
		bucket <<= 1
	}
	return bucket
}

// PadPayload creates a padded payload buffer of the appropriate bucket size
// containing uint16(len(plaintext)) || plaintext || zero-padding.
func PadPayload(plaintext []byte) ([]byte, error) {
	if len(plaintext) > u16Max-2 {
		return nil, fmt.Errorf("payload length %d exceeds max %d", len(plaintext), u16Max-2)
	}
	bucket := PadBucketSize(len(plaintext))
	if len(plaintext)+2 > bucket {
		return nil, fmt.Errorf("payload %d + prefix exceeds bucket size %d", len(plaintext), bucket)
	}
	buf := make([]byte, bucket)
	binary.BigEndian.PutUint16(buf[:2], uint16(len(plaintext)))
	copy(buf[2:], plaintext)
	return buf, nil
}

// UnpadPayload extracts the unpadded plaintext from a padded payload buffer, verifying
// the prefix length and ensuring all trailing padding bytes are strictly zero.
func UnpadPayload(padded []byte) ([]byte, error) {
	if len(padded) < 2 {
		return nil, fmt.Errorf("%w: padded buffer too short (%d bytes)", ErrMalformedFrame, len(padded))
	}
	unpaddedLen := int(binary.BigEndian.Uint16(padded[:2]))
	if 2+unpaddedLen > len(padded) {
		return nil, fmt.Errorf("%w: unpadded length %d exceeds buffer %d", ErrMalformedFrame, unpaddedLen, len(padded))
	}
	for i, b := range padded[2+unpaddedLen:] {
		if b != 0 {
			return nil, fmt.Errorf("%w: non-zero padding byte at index %d", ErrMalformedFrame, 2+unpaddedLen+i)
		}
	}
	return padded[2 : 2+unpaddedLen], nil
}

// encodeFrameHeader encodes h into a fresh 16-byte buffer.
func encodeFrameHeader(h FrameHeader) []byte {
	buf := make([]byte, frameHeaderBytes)
	buf[0] = h.Version
	buf[1] = h.Type
	buf[2] = h.Flags
	buf[3] = 0 // reserved
	binary.BigEndian.PutUint16(buf[4:6], h.FileIdx)
	binary.BigEndian.PutUint32(buf[6:10], h.BlockIdx)
	binary.BigEndian.PutUint32(buf[10:14], h.FrameOff)
	binary.BigEndian.PutUint16(buf[14:16], h.Len)
	return buf
}

// decodeFrameHeader decodes a header from the first 16 bytes of buf.
func decodeFrameHeader(buf []byte) (FrameHeader, error) {
	if len(buf) < frameHeaderBytes {
		return FrameHeader{}, fmt.Errorf("frame header needs %d bytes, got %d", frameHeaderBytes, len(buf))
	}
	return FrameHeader{
		Version:  buf[0],
		Type:     buf[1],
		Flags:    buf[2],
		FileIdx:  binary.BigEndian.Uint16(buf[4:6]),
		BlockIdx: binary.BigEndian.Uint32(buf[6:10]),
		FrameOff: binary.BigEndian.Uint32(buf[10:14]),
		Len:      binary.BigEndian.Uint16(buf[14:16]),
	}, nil
}

File: packages/protocol/src/frame.ts
/**
 * Binary codec for the 16-byte frame header. The encoded header bytes
 * are used verbatim as the AES-GCM AAD, so encode/decode must be exact and stable.
 *
 * Layout (big-endian):
 *   version(u8) type(u8) flags(u8) reserved(u8)
 *   fileIdx(u16) blockIdx(u32) frameOff(u32) len(u16)
 *
 * frameOff is a byte offset within the block, so it must span the whole block: at the
 * default 1 MiB block a 16 KiB frame reaches offset ~1 MiB, far past u16. It is u32
 * (an earlier draft miscalculated it as u16), absorbed by dropping the former trailing
 * reserved(u16) — the header stays a fixed 16 bytes.
 */

import { FRAME_HEADER_BYTES, MAX_PAD_BUCKET_BYTES, MIN_PAD_BUCKET_BYTES } from './constants.js';
import type { FrameHeader } from './transfer.js';

const U8_MAX = 0xff;
const U16_MAX = 0xffff;
const U32_MAX = 0xffffffff;

function assertRange(name: string, value: number, max: number): void {
  if (!Number.isInteger(value) || value < 0 || value > max) {
    throw new RangeError(`frame header field ${name}=${value} out of range [0, ${max}]`);
  }
}

/** Encode a header into a fresh 16-byte buffer. */
export function encodeFrameHeader(h: FrameHeader): Uint8Array {
  assertRange('version', h.version, U8_MAX);
  assertRange('type', h.type, U8_MAX);
  assertRange('flags', h.flags, U8_MAX);
  assertRange('fileIdx', h.fileIdx, U16_MAX);
  assertRange('blockIdx', h.blockIdx, U32_MAX);
  assertRange('frameOff', h.frameOff, U32_MAX);
  assertRange('len', h.len, U16_MAX);

  const buf = new Uint8Array(FRAME_HEADER_BYTES);
  const dv = new DataView(buf.buffer);
  dv.setUint8(0, h.version);
  dv.setUint8(1, h.type);
  dv.setUint8(2, h.flags);
  dv.setUint8(3, 0); // reserved
  dv.setUint16(4, h.fileIdx, false);
  dv.setUint32(6, h.blockIdx, false);
  dv.setUint32(10, h.frameOff, false);
  dv.setUint16(14, h.len, false);
  return buf;
}

/** Decode a header from the first 16 bytes of `buf`. */
export function decodeFrameHeader(buf: Uint8Array): FrameHeader {
  if (buf.byteLength < FRAME_HEADER_BYTES) {
    throw new RangeError(`frame header needs ${FRAME_HEADER_BYTES} bytes, got ${buf.byteLength}`);
  }
  const dv = new DataView(buf.buffer, buf.byteOffset, FRAME_HEADER_BYTES);
  return {
    version: dv.getUint8(0),
    type: dv.getUint8(1),
    flags: dv.getUint8(2),
    fileIdx: dv.getUint16(4, false),
    blockIdx: dv.getUint32(6, false),
    frameOff: dv.getUint32(10, false),
    len: dv.getUint16(14, false),
  };
}

/**
 * Returns the smallest power-of-two bucket size >= (unpaddedLen + 2),
 * clamped to [MIN_PAD_BUCKET_BYTES, MAX_PAD_BUCKET_BYTES].
 */
export function padBucketSize(unpaddedLen: number): number {
  const target = unpaddedLen + 2;
  if (target <= MIN_PAD_BUCKET_BYTES) {
    return MIN_PAD_BUCKET_BYTES;
  }
  if (target > 32768) {
    return MAX_PAD_BUCKET_BYTES;
  }
  let bucket = MIN_PAD_BUCKET_BYTES;
  while (bucket < target) {
    bucket <<= 1;
  }
  return bucket;
}

/**
 * Creates a padded payload buffer of the appropriate bucket size containing
 * `uint16(unpaddedLen) || plaintext || zero-padding`.
 */
export function padPayload(plaintext: Uint8Array): Uint8Array {
  if (plaintext.length > U16_MAX - 2) {
    throw new RangeError(`payload length ${plaintext.length} exceeds max ${U16_MAX - 2}`);
  }
  const bucket = padBucketSize(plaintext.length);
  if (plaintext.length + 2 > bucket) {
    throw new RangeError(`payload ${plaintext.length} + prefix exceeds bucket size ${bucket}`);
  }
  const buf = new Uint8Array(bucket);
  new DataView(buf.buffer, buf.byteOffset, 2).setUint16(0, plaintext.length, false);
  buf.set(plaintext, 2);
  return buf;
}

/**
 * Extracts unpadded plaintext from a padded payload buffer, verifying the prefix
 * length and ensuring all trailing padding bytes are strictly zero.
 */
export function unpadPayload(padded: Uint8Array): Uint8Array {
  if (padded.length < 2) {
    throw new Error(`malformed frame: padded buffer too short (${padded.length} bytes)`);
  }
  const unpaddedLen = new DataView(padded.buffer, padded.byteOffset, 2).getUint16(0, false);
  if (2 + unpaddedLen > padded.length) {
    throw new Error(
      `malformed frame: unpadded length ${unpaddedLen} exceeds buffer ${padded.length}`,
    );
  }
  for (let i = 2 + unpaddedLen; i < padded.length; i++) {
    if (padded[i] !== 0) {
      throw new Error(`malformed frame: non-zero padding byte at index ${i}`);
    }
  }
  return padded.subarray(2, 2 + unpaddedLen);
}
